### ISSUE TITLE
MPI_Scatter in parallel.h?

### ISSUE THREAD

**Issue by permcody** [libMesh/libmesh#1058]
In doing some preliminary runs of the optimized grain tracker I made crude observations of the memory usage and am simply blown away by how much my data structures appear to occupy. Running 6000 grains on 180^3 mesh seemed to increase the memory usage right after the `allgather` operation (cause of the non-issue in #1039) by over a GB per processor! This is the optimized version mind you!  The back of the envelope calculations don't seem to add up to this amount of memory but I do have a lot of data in sets which takes often more than 4x the amount of memory for POD types AND that data structure exists on every processor, so maybe...

I'm thinking about really digging into this algorithm and changing it so that only the root process maintains the big data structure and the remaining ranks maintain only local data and local to global mappings. I don't think this will be too difficult but I could really use `MPI_Scatter` to get this done. Can we add this to parallel.h?

TL;DR;
I'd love to have MPI_Scatter in parallel.h. Any objections?

**Comment by roystgnr** [comment]
Sounds good to me!

**Comment by permcody** [comment]
I'm starting to flesh out the `scatter` operations. Before I finish this up I want to make sure that everyone (@roystgnr, @jwpeterson, @friedmud, etc) like the API I'm setting up. Here are a few options:

```
/**
 * Takes a vector equal to the communicator size and sends the ith item to the ith processor
 * stores it in recv.
 */
template <typename T>
inline void Communicator::scatter (std::vector<T> & data, T & recv, const unsigned int root_id=0) const;

/**
  * Take a "stacked" vector and distributes equal sized chunks to each processor and 
  * stores in recv. The recv vectors do not have to be sized for this operation.
  */
template <typename T>
inline void Communicator::scatter (std::vector<T> & data, std::vector<T> & recv, const unsigned int root_id=0) const;

/**
  * Takes a "stacked" data vector and a vector of sizes for distributing non-uniform 
  * sized chunks to all processors.
  * The recv vectors do not have to be sized for this operation.
  */
template <typename T>
inline void Communicator::scatter (std::vector<T> & data, std::vector<std::size_t> counts, std::vector<T> & recv, const unsigned int root_id=0) const;
```

**Comment by jwpeterson** [comment]
This seems fine to me...

**Comment by roystgnr** [comment]
This all looks good exactly as-is.

Using size_t instead of int or unsigned int is questionable, though - with size_t you'll need to copy to a temporary buffer before invoking MPI_Scatterv, whereas with int or unsigned int you'll just need a cast.

I personally do prefer the forward-compatible API with size_t over the infintesimally faster API with int, but just wanted to bring up the option explicitly.

**Comment by permcody** [comment]
Closed by #1062

**Comment by permcody** [comment]
I'm thinking about adding another "overload" already. This version would assert that the outer vector is equal to the size of the communicator then build the offsets and counts vectors and scatter the inner (possibly jagged) vectors to the respective processors. Any objections?
```
template <typename T>
void Communicator::scatter(const std::vector<std::vector<T> > & data,
                           std::vector<T> & recv,
                           const unsigned int root_id) const;
```

**Comment by friedmud** [comment]
Makes sense

**Comment by roystgnr** [comment]
I really like that, yeah.
